### PR TITLE
Implemented Tests for Carousel, Image, Text and Removed Magical Numbers

### LINKED ISSUE #93: Test Carousel, TextStep, ImageStep, Dialogue, Message, Scorable
Test Carousel, TextStep, ImageStep, Dialogue, Message, Scorable classes

### PR DESCRIPTION
**Pull request by ZZBP**
100% line coverage for  Carousel, Image, Text entities;
Tentatively created TalkStepTests so that the step tests can extend;
Created ScoreStandards to store our scaling(keywords weighting, max score weighting); 
Changed the step entities according to the ScoreStandards.

### CODE REVIEW

File: backend/src/test/java/com/speakfluid/backend/entities/DialogueTests.java [hunk 1]
```diff
@@ -1,19 +1,59 @@
...
+ * @author  Kai Zhuang, Zoey Zhang
+ * @version 2.0
+ * @since   2022-11-26
  */
 
 class DialogueTests {
+    static Dialogue<WozMessage> d1;
+    static ArrayList<WozMessage> botMsg1;
+    static ArrayList<WozMessage> userMsg1;
+    static WozMessage m1 = new com.speakfluid.backend.entities.WozMessage("text", "Hi, here are the locations of our stores");
```
Comment (Lemiex) [inline backend/src/test/java/com/speakfluid/backend/entities/DialogueTests.java]: Wondering why it's `com.speakfluid.backend.entities.WozMessage` instead of just `WozMessage`? You imported all the entities on the top do you should be able to make it just WozMessage

Comment (Lemiex) [reply backend/src/test/java/com/speakfluid/backend/entities/DialogueTests.java]: Also, the trace_type for these WozMessages should be 'request' and 'response'

File: backend/src/test/java/com/speakfluid/backend/entities/DialogueTests.java [hunk 2]
```diff
@@ -1,19 +1,59 @@
...
+    static WozMessage m2 = new com.speakfluid.backend.entities.WozMessage("text", "What else can I help you with today?");
+    static WozMessage m3 = new WozMessage("text", "Hi,");
+    static WozMessage m4 = new WozMessage("text", "Can I have the direction to the store?");
+    @BeforeAll
+    public static void setUp(){
+        botMsg1 = new ArrayList<>(Arrays.asList(m1,m2));
+        userMsg1 = new ArrayList<>(Arrays.asList(m3, m4));
```
Comment (Lemiex) [inline backend/src/test/java/com/speakfluid/backend/entities/DialogueTests.java:33]: I like how you are testing when chatbotMsg and userMsg have more than one WozMessage

File: backend/src/test/java/com/speakfluid/backend/entities/DialogueTests.java [hunk 3]
```diff
@@ -1,19 +1,59 @@
...
+        assertEquals(botMsg1, d1.getChatBotMessage());
+
+    }
+    @Test
+    void testGetUserMessage() {
+        assertEquals(userMsg1, d1.getUserMessage());
+
+    }
+    @Test
+    void testGetStepScorePairs() {
```
Comment (Lemiex) [inline backend/src/test/java/com/speakfluid/backend/entities/DialogueTests.java]: I'm a bit confused what this test does, what is step score pairs?

File: backend/src/test/java/com/speakfluid/backend/entities/ImageStepTests.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    static ImageStep imageStep = new ImageStep();
+    static Dialogue<WozMessage> d1;
+    static Dialogue<WozMessage> d2;
+    static Dialogue<WozMessage> d3;
+    static ArrayList<WozMessage> botMsg1;
+    static ArrayList<WozMessage> userMsg1;
+    static ArrayList<WozMessage> botMsg2;
+    static ArrayList<WozMessage> userMsg2;
+    static ArrayList<WozMessage> userMsg3;
+    static WozMessage m1 = new WozMessage("text", "Hi, here are the locations of our stores");
```
Comment (Lemiex) [inline backend/src/test/java/com/speakfluid/backend/entities/ImageStepTests.java]: change trace type

File: backend/src/test/java/com/speakfluid/backend/entities/ImageStepTests.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    }
+
+    @Test
+    void testGetMaxScore() {
+        assertEquals(18.0, imageStep.getMaxScore());
+    }
+
+    @Test
+    void testGetScoreAccumulator() {
+        assertEquals(26.0, imageStep.getScoreAccumulator());
```
Comment (Lemiex) [inline backend/src/test/java/com/speakfluid/backend/entities/ImageStepTests.java]: why would the scoreAccumulator be 26.0 here? no runAnalysis() is run right

File: backend/src/test/java/com/speakfluid/backend/entities/ImageStepTests.java [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+        assertEquals(26.0, actualScore);
+    }
+    @Test
+    void testRunAnalysisWithBotMsgMatch() {
+        imageStep.runAnalysis(d3);
+        double actualScore = imageStep.getScoreAccumulator();
+        assertEquals(9.0, actualScore);
+    }
+    @Test
+    void testRunAnalysisWithUserMsgEmpty() {
```
Comment (Lemiex) [inline backend/src/test/java/com/speakfluid/backend/entities/ImageStepTests.java:87]: awesome testing edge cases!

File: backend/src/main/java/com/speakfluid/backend/entities/CarouselStep.java [hunk 7]
```diff
@@ -22,44 +22,38 @@
...
-    private final int maxScore = 16;
-    private final String stepName = "Carousel";
+    private double scoreAccumulator;
+    private final double maxScore = ScoreStandards.standardStepClass;
     private final List<Map<String, Double>> imageKeyWords = Arrays.asList(
-            Map.ofEntries(entry("here are", 2.0),entry("map", 1.0), entry("location", 1.0),
-                    entry("direction", 1.0)),
-            Map.ofEntries(entry("here is a picture", 4.0), entry("here is an image", 4.0),
-                    entry("picture", 3.0),entry("illustration", 3.0),entry("image", 3.0)
+            Map.ofEntries(entry("here are", ScoreStandards.mediumMatch),entry("map", ScoreStandards.lowMatch),
```
Comment (Lemiex) [inline backend/src/main/java/com/speakfluid/backend/entities/CarouselStep.java:28]: Thanks for getting rid of magic numbers!!

File: backend/src/main/java/com/speakfluid/backend/entities/TalkStep.java [hunk 8]
```diff
@@ -81,6 +81,6 @@ public int countMatchKeywords(Message message, List<Map<String, Double>> keyword
     //     return minutes*60 + seconds;
     // }
 
-    public abstract void runAnalysis(Dialogue dialogue);
+    public abstract void runAnalysis(Dialogue<?> dialogue);
```
Comment (Lemiex) [inline backend/src/main/java/com/speakfluid/backend/entities/TalkStep.java:84]: gets rid of 'raw use of parameterized variable'

**Review by Lemiex — APPROVED** [review]
Amazing Zoey!

File: backend/src/test/java/com/speakfluid/backend/entities/DialogueTests.java (hunk 1 above)
Comment (ZZBP) [reply backend/src/test/java/com/speakfluid/backend/entities/DialogueTests.java]: I will change the trace type, the test for Dialogue still needs some changes and I will let you know when the final version is up!!! Thank you so so much for the throughout review!!!!!!

File: backend/src/test/java/com/speakfluid/backend/entities/ImageStepTests.java (hunk 5 above)
Comment (ZZBP) [reply backend/src/test/java/com/speakfluid/backend/entities/ImageStepTests.java]: That is an issue I am tackling, somehow the scoreaccumulator was carried over from other tests.

File: backend/src/test/java/com/speakfluid/backend/entities/DialogueTests.java (hunk 3 above)
Comment (ZZBP) [reply backend/src/test/java/com/speakfluid/backend/entities/DialogueTests.java]: This is no longer tested!

File: backend/src/test/java/com/speakfluid/backend/entities/DialogueTests.java (hunk 1 above)
Comment (ZZBP) [reply backend/src/test/java/com/speakfluid/backend/entities/DialogueTests.java]: It's updated!

**Review by auroraborie — COMMENTED** [review]
Hey Zoey wow thanks for the massive PR and standardizing our scores across TalkSteps!
Just a note on the testing, Ekagra (the TA) suggested we use mock values in our tests so we don't end up getting the wrong value and not knowing if it is because of the specific TalkStep class or its parent. I suggest you implement this in the runAnalysis functions tests. The syntax would be: TalkStep talkstep = mock(TalkStep.class);
        Mockito.when(talkstep.calculateMsgLength(user3)).thenReturn(3);

I think this would allow for more accurate testing and is a small change.

**Comment by Lemiex** [comment]
Hey rory I noticed that as well, after I suggested it to you and when to write it in my tests, I realised it doesn't actually work ;( Which is why it's not in my PR's tests yet either. I got rid of it for now and the tests pass but we'll need to return to this later

**Review by nm-le — APPROVED** [review]

**Comment by auroraborie** [comment]
> Hey rory I noticed that as well, after I suggested it to you and when to write it in my tests, I realised it doesn't actually work ;( Which is why it's not in my PR's tests yet either. I got rid of it for now and the tests pass but we'll need to return to this later

**Comment by auroraborie** [comment]
> Hey rory I noticed that as well, after I suggested it to you and when to write it in my tests, I realised it doesn't actually work ;( Which is why it's not in my PR's tests yet either. I got rid of it for now and the tests pass but we'll need to return to this later

okay i see then i have also made the changes in my tests :)

**Comment by ZZBP** [comment]
Hey Rory, thank you for addressing the mocking issue! We will need to consult further with the TA for a solution! We will first merge this pr so that ScoreStandards will be merged to main, and people can access it! We can easily add Mockito afterward!
